services/seats.py

```diff
@@ -0,0 +1,118 @@
+import logging
+from dataclasses import dataclass
+from enum import Enum
+
+from sqlalchemy import func
+from sqlalchemy.orm import Session
+
+from database.models import Owner
+from services.billing import is_pr_billing_plan
+from services.decoration import _is_bot_account
+from services.repository import EnrichedPull
+
+log = logging.getLogger(__name__)
+
+
+class ShouldActivateSeat(Enum):
+    NO_ACTIVATE = "no_activate"
+    MANUAL_ACTIVATE = "manual_activate"
+    AUTO_ACTIVATE = "auto_activate"
+
+
+@dataclass
+class SeatActivationInfo:
+    should_activate_seat: ShouldActivateSeat = ShouldActivateSeat.NO_ACTIVATE
+    owner_id: int | None = None
+    author_id: int | None = None
+    reason: str | None = None
+
+
+def determine_seat_activation(pull: EnrichedPull) -> SeatActivationInfo:
+    """
+    this function will determine if a user needs to be activated based on information about the user, their org, their repo, and the PR
+    1. if the repo is public they don't need to be activated
+    2. get repostiory owner info
+    2.1 (custom gitlab logic) if they are on gitlab we get their root group as the org instead of the repo owner
+    3. get pr author info
+    4. if org doesn't use seats, or author is included in seats already then no need to be activated
+    5. if author is bot user then no need to be activated
+    6. user must either be manually activated or auto activated
+    """
+    db_pull = pull.database_pull
+    provider_pull = pull.provider_pull
+    if provider_pull is None:
+        log.warning(
+            "Provider pull was None when determining whether to activate seat for user",
+            extra=dict(
+                pullid=db_pull.pullid,
+                repoid=db_pull.repoid,
+                head_commit=db_pull.head,
+                base_commit=db_pull.base,
+            ),
+        )
+        return SeatActivationInfo(reason="no_provider_pull")
+
+    if db_pull.repository.private is False:
+        return SeatActivationInfo(reason="public_repo")
+
+    org = db_pull.repository.owner
+
+    db_session: Session = db_pull.get_db_session()
+    if org.service == "gitlab" and org.parent_service_id:
+        # need to get root group so we can check plan info
+        (gl_root_group,) = db_session.query(
```

Comment:
This returns a tuple?

Reply:
i ripped this code straight from the existing `services.decoration` file, not too sure on the context: https://github.com/codecov/worker/blob/main/services/decoration.py#L99

Reply:
mmm, Ill assume it's fine but we can keep an eye on this